ui/src/logs/actions/index.ts

```diff
@@ -0,0 +1,83 @@
+import {Source, Namespace, TimeRange} from 'src/types'
+import {getSource} from 'src/shared/apis'
+import {getDatabasesWithRetentionPolicies} from 'src/shared/apis/databases'
+
+export enum ActionTypes {
+  SetSource = 'LOGS_SET_SOURCE',
+  SetNamespaces = 'LOGS_SET_NAMESPACES',
+  SetTimeRange = 'LOGS_SET_TIMERANGE',
+  SetNamespace = 'LOGS_SET_NAMESPACE',
+}
+
+interface SetSourceAction {
+  type: ActionTypes.SetSource
+  payload: {
+    source: Source
+  }
+}
+
+interface SetNamespacesAction {
+  type: ActionTypes.SetNamespaces
+  payload: {
+    namespaces: Namespace[]
+  }
+}
+
+interface SetNamespaceAction {
+  type: ActionTypes.SetNamespace
+  payload: {
+    namespace: Namespace
+  }
+}
+
+interface SetTimeRangeAction {
+  type: ActionTypes.SetTimeRange
+  payload: {
+    timeRange: TimeRange
+  }
+}
+
+export type Action =
+  | SetSourceAction
+  | SetNamespacesAction
+  | SetTimeRangeAction
+  | SetNamespaceAction
+
+export const setSource = (source: Source): SetSourceAction => ({
+  type: ActionTypes.SetSource,
+  payload: {
+    source,
+  },
+})
+
+export const setNamespace = (namespace: Namespace): SetNamespaceAction => ({
+  type: ActionTypes.SetNamespace,
+  payload: {
+    namespace,
+  },
+})
+
+export const setNamespaces = (
+  namespaces: Namespace[]
+): SetNamespacesAction => ({
+  type: ActionTypes.SetNamespaces,
+  payload: {
+    namespaces,
+  },
+})
+
+export const setTimeRange = (timeRange: TimeRange): SetTimeRangeAction => ({
+  type: ActionTypes.SetTimeRange,
+  payload: {
+    timeRange,
+  },
+})
+
+export const getSourceAsync = sourceID => async dispatch => {
+  const response = await getSource(sourceID)
+  const source = response.data
+  const namespaces = await getDatabasesWithRetentionPolicies(source.links.proxy)
```

Comment:
1. it does seem like in some cases, deep dot access is ok -- thoughts on this?
1. what do you think about either factoring this second promise out into a separate function, or renaming this current function to something that implies that it gets both sources and databases with retention policies?

Reply:
This happens when I select a source I need to get that source and since I have changed sources I need to fetch all databases for that source. 👍 to fixing the nested lookup.